### PR TITLE
Fifth task from Threads

### PR DESCRIPTION
**Pull request by p-nedelchev**
Needs Review.

### CODE REVIEW

File: src/main/java/com/clouway/fifth/TimeoutHashtable.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package com.clouway.fifth;
+
+import java.util.*;
+
+/**
+ * @author Petar Nedelchev <peter.krasimirov@gmail.com>
+ */
+public class TimeoutHashtable<K, V> {
+    private final Hashtable<K,V> table;
+    private Map<K, Long> keyUsage;
```
Comment (tsony-tsonev) [inline src/main/java/com/clouway/fifth/TimeoutHashtable.java]: You can use set here

File: src/main/java/com/clouway/fifth/TimeoutHashtable.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            }
+            return null;
+        }
+    }
+
+    public V remove(K key) {
+        synchronized (table) {
+            if (table.containsKey(key)) {
+                V value = table.get(key);
+                table.remove(key);
```
Comment (tsony-tsonev) [inline src/main/java/com/clouway/fifth/TimeoutHashtable.java]: you should also remove from the keyUsage map, cause it is gonna get flooded with keys, which is memory leak

File: src/main/java/com/clouway/fifth/TimeoutHashtable.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    public void removeExpired() {
+           Iterator<K> iter = keyUsage.keySet().iterator();
+           while(iter.hasNext()) {
+                K key = iter.next();
+                if (isExpired(key)) {
+                    synchronized (table) {
+                        System.out.println("Key: " + key + " expired. Removed");
+                        remove(key);
+                        iter.remove();
+                        table.notifyAll();
```
Comment (tsony-tsonev) [inline src/main/java/com/clouway/fifth/TimeoutHashtable.java]: not sure this notify is needed, also you should break in the else, cause you don't want to traverse through all the keys which are known not to be expired

File: src/main/java/com/clouway/fifth/Demo.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        final Random generator = new Random();
+        Thread addingThread = new Thread(new Runnable() {
+            public void run() {
+                while (true) {
+                    int randomKey = generator.nextInt(100);
+                    int randomValue = generator.nextInt(100);
+
+                    try {
+                        table.put(randomKey, randomValue);
+                        Thread.sleep(800);
```
Comment (p-nedelchev) [inline src/main/java/com/clouway/fifth/Demo.java]: I simplified it.

File: src/main/java/com/clouway/fifth/Demo.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                }
+
+            }
+        });
+        Thread checkingThread = new Thread(new Runnable() {
+            public void run() {
+                while (true) {
+                    try {
+                        Thread.sleep(1000);
+                        table.removeExpired();
```
Comment (p-nedelchev) [inline src/main/java/com/clouway/fifth/Demo.java]: Thread is in the class implementation now.

File: src/main/java/com/clouway/fifth/TimeoutHashtable.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        keyUsage = new LinkedHashMap<K, Long>();
+        this.timeout = timeout;
+    }
+
+    public void put(K key, V value) {
+        synchronized (table) {
+            table.put(key, value);
+            keyUsage.put(key, System.currentTimeMillis());
+            System.out.println(Thread.currentThread().getName() + " added " + "(" + key + ", " + value + ")");
+            table.notifyAll();
```
Comment (p-nedelchev) [inline src/main/java/com/clouway/fifth/TimeoutHashtable.java]: I removed it still works.

File: src/main/java/com/clouway/fifth/TimeoutHashtable.java (hunk 2 above)
Comment (p-nedelchev) [inline src/main/java/com/clouway/fifth/TimeoutHashtable.java]: The thread that observes the keys in keyUsage deletes them so there is no memory leak.

File: src/main/java/com/clouway/fifth/TimeoutHashtable.java (hunk 3 above)
Comment (p-nedelchev) [inline src/main/java/com/clouway/fifth/TimeoutHashtable.java]: Done. I added else to break the iteration of the map when there are no more keys that are expired.

**Comment by p-nedelchev** [comment]
All fixes applied.

**Comment by tsony-tsonev** [comment]
LGTM

**Pull request merged**